### PR TITLE
Adding some utility functions to P4 and other classes

### PR DESCRIPTION
**Pull request by yiiyama**
Adding functions to BareCollection, BareP4 and derived classes for ease of use when running analysis on Nero files using these classes.
 - virtual unsigned size()
 - virtual void init() to new pointer members
 - setBranchAddresses() only picks branches with nonzero status
 - Proposing convention to zero out pointer members in the class definition (allowed since C++11) to avoid the common error of uninitialized pointers.

### CODE REVIEW

File: Core/src/BareFatJets.cpp [hunk 1]
```diff
@@ -3,24 +3,54 @@
...
-    trimmedMass = NULL;
-    prunedMass = NULL;
-    filteredMass = NULL;
-    softdropMass = NULL;
-    ak8_subjet = NULL;
-    ak8jet_hasSubjet = NULL;
-    ak8subjet_btag = NULL;
-    hbb = NULL;
-
 }
```
Comment (amarini) [inline Core/src/BareFatJets.cpp]: If I didn't init, the destructor will crash ?

File: Core/src/BareMonteCarlo.cpp [hunk 2]
```diff
@@ -52,21 +58,31 @@ void BareMonteCarlo::defineBranches(TTree *t){
...
+    if (t->GetBranchStatus("pdfX1"))
+        t->SetBranchAddress("pdfX1"	,&x1		);
+    if (t->GetBranchStatus("pdfX2"))
+        t->SetBranchAddress("pdfX2"	,&x2		);
+    if (t->GetBranchStatus("pdfId1"))
+        t->SetBranchAddress("pdfId1"	,&pdf1Id	);
+    if (t->GetBranchStatus("pdfId2"))
+        t->SetBranchAddress("pdfId2"	,&pdf2Id	);
+    if (t->GetBranchStatus("pdfScalePdf"))
+        t->SetBranchAddress("pdfScalePdf"	,&scalePdf	);
```
Comment (amarini) [inline Core/src/BareMonteCarlo.cpp]: Maybe the if(t_>...) t->Set ... and the 
init (if i !=NULL ) ... can be templated.
For the moment I'm fine.

Comment (yiiyama) [inline Core/src/BareMonteCarlo.cpp]: You mean defining a compiler macro? Yeah that's a way to do it.

File: Core/src/BareFatJets.cpp (hunk 1 above)
Comment (yiiyama) [inline Core/src/BareFatJets.cpp]: No, delete will do nothing if the pointer is properly zero initialized. The initialization is done in the .h file.

File: Core/src/BareMonteCarlo.cpp (hunk 2 above)
Comment (amarini) [inline Core/src/BareMonteCarlo.cpp]: Also a template function ?

BareFunctions.hpp

template <class T>
SetAddress( TTree *t , string name , T &x)
{
 if (t->GetBranchStatus( name.c_str())
 t->SetBranchAddress( name.c_str(), &x) ;
return ;
}

Andrea

Il 06.09.15 23:03, Yutaro Iiyama ha scritto:
>
> In Core/src/BareMonteCarlo.cpp
> <https://github.com/MiT-HEP/NeroProducer/pull/106#discussion_r38826087>:
>
> > +    if (t->GetBranchStatus("pdfQscale"))
> > +        t->SetBranchAddress("pdfQscale"	,&qScale	);
> > +    if (t->GetBranchStatus("pdfAlphaQED"))
> > +        t->SetBranchAddress("pdfAlphaQED"	,&alphaQED	);
> > +    if (t->GetBranchStatus("pdfAlphaQCD"))
> > +        t->SetBranchAddress("pdfAlphaQCD"	,&alphaQCD	);
> > +    if (t->GetBranchStatus("pdfX1"))
> > +        t->SetBranchAddress("pdfX1"	,&x1		);
> > +    if (t->GetBranchStatus("pdfX2"))
> > +        t->SetBranchAddress("pdfX2"	,&x2		);
> > +    if (t->GetBranchStatus("pdfId1"))
> > +        t->SetBranchAddress("pdfId1"	,&pdf1Id	);
> > +    if (t->GetBranchStatus("pdfId2"))
> > +        t->SetBranchAddress("pdfId2"	,&pdf2Id	);
> > +    if (t->GetBranchStatus("pdfScalePdf"))
> > +        t->SetBranchAddress("pdfScalePdf"	,&scalePdf	);
>
> You mean defining a compiler macro? Yeah that's a way to do it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/MiT-HEP/NeroProducer/pull/106/files#r38826087>.
>


-- 
●●●●●●●●●●●●●●●●●●●●●●●●●●
Andrea Carlo Marini, 
MIT Post-Doc,
c/o 32-4-A20 CERN

File: Core/src/BareFatJets.cpp (hunk 1 above)
Comment (amarini) [inline Core/src/BareFatJets.cpp]: Il 06.09.15 23:03, Yutaro Iiyama ha scritto:
>
> In Core/src/BareFatJets.cpp
> <https://github.com/MiT-HEP/NeroProducer/pull/106#discussion_r38826111>:
>
> > @@ -3,24 +3,54 @@
> >  
> >  
> >  BareFatJets::BareFatJets(): BareP4(){
> > -    p4 = NULL;
> > -    rawPt = NULL;
> > -    flavour = NULL;
> > -    tau1 = NULL;
> > -    tau2 = NULL;
> > -    tau3 = NULL;
> > -    trimmedMass = NULL;
> > -    prunedMass = NULL;
> > -    filteredMass = NULL;
> > -    softdropMass = NULL;
> > -    ak8_subjet = NULL;
> > -    ak8jet_hasSubjet = NULL;
> > -    ak8subjet_btag = NULL;
> > -    hbb = NULL;
> > -
> >  }
> >  
>
> No, delete will do nothing if the pointer is properly zero
> initialized. The initialization is done in the .h file.
>
Is this a new in C++11 ... awsome!


Andrea
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/MiT-HEP/NeroProducer/pull/106/files#r38826111>.
>


-- 
●●●●●●●●●●●●●●●●●●●●●●●●●●
Andrea Carlo Marini, 
MIT Post-Doc,
c/o 32-4-A20 CERN

**Pull request merged**